File: mobile/lib/core/srs/fsrs_engine.dart
/// Moteur FSRS-5 — cœur scientifique de MedAnki DZ.
///
/// Toutes les fonctions de ce fichier sont **pures** : elles ne lisent aucune
/// horloge, aucun stockage, aucun réseau. L'instant courant est toujours passé
/// en paramètre (`nowMs`). C'est ce qui rend le moteur testable de façon
/// déterministe et rejouable — exigence non négociable de l'architecture v2 :
///
///     srs_state = FSRS.fold(events.where(!examMode).sortBy(reviewedAt))
///
/// Les formules sont alignées sur `ts-fsrs` (backend, Phase 6) afin que les
/// deux implémentations produisent des états identiques.
library;

import 'dart:math' as math;

import 'fsrs_parameters.dart';
import 'review_event.dart';
import 'srs_models.dart';

class FsrsEngine {
  const FsrsEngine({this.parameters = const FsrsParameters()});

  final FsrsParameters parameters;

  // --------------------------------------------------------------------------
  // Primitives du modèle
  // --------------------------------------------------------------------------

  /// Probabilité de rappel après [elapsedDays] jours : R(t,S) = (1 + F·t/S)^D.
  ///
  /// Vaut 1 à t = 0 et 0.9 lorsque t = S (par construction de [kFactor]).
  double retrievability(double elapsedDays, double stability) {
    if (stability <= 0) return 0.0;
    final double t = math.max(0.0, elapsedDays);
    return math.pow(1.0 + kFactor * t / stability, kDecay).toDouble();
  }

  /// Intervalle en jours pour atteindre la rétention cible depuis [stability].
  double intervalFromStability(double stability) {
    final double r = parameters.requestRetention;
    return (stability / kFactor) * (math.pow(r, 1.0 / kDecay) - 1.0);
  }

  /// Stabilité initiale, fonction de la première note.
  double _initialStability(Rating rating) {
    return _clamp(
        parameters.w(rating.value - 1), kMinStability, kMaxStability);
  }

  /// Difficulté initiale : D0(G) = w4 - e^(w5·(G-1)) + 1.
  double _initialDifficulty(Rating rating) {
    final double d =
        parameters.w(4) - math.exp(parameters.w(5) * (rating.value - 1)) + 1.0;
    return _clamp(d, kMinDifficulty, kMaxDifficulty);
  }

  /// Amortissement linéaire : les cartes déjà difficiles bougent moins.
  double _linearDamping(double deltaD, double difficulty) {
    return deltaD * (10.0 - difficulty) / 9.0;
  }

  /// Retour à la moyenne vers la difficulté d'une carte notée "Easy".
  double _meanReversion(double init, double current) {
    return parameters.w(7) * init + (1.0 - parameters.w(7)) * current;
  }

  /// Difficulté après une revue notée [rating].
  double _nextDifficulty(double difficulty, Rating rating) {
    final double deltaD = -parameters.w(6) * (rating.value - 3);
    final double damped = difficulty + _linearDamping(deltaD, difficulty);
    final double reverted =
        _meanReversion(_initialDifficulty(Rating.easy), damped);
    return _clamp(reverted, kMinDifficulty, kMaxDifficulty);
  }

  /// Stabilité après un rappel réussi (Hard / Good / Easy).
  ///
  /// Le gain est d'autant plus fort que la carte était sur le point d'être
  /// oubliée (terme `e^((1-R)·w10)`) : réviser trop tôt n'apporte presque rien.
  double _nextRecallStability(
      double d, double s, double r, Rating rating) {
    final double hardPenalty =
        rating == Rating.hard ? parameters.w(15) : 1.0;
    final double easyBonus = rating == Rating.easy ? parameters.w(16) : 1.0;
    final double next = s *
        (1.0 +
            math.exp(parameters.w(8)) *
                (11.0 - d) *
                math.pow(s, -parameters.w(9)) *
                (math.exp((1.0 - r) * parameters.w(10)) - 1.0) *
                hardPenalty *
                easyBonus);
    return _clamp(next, kMinStability, kMaxStability);
  }

  /// Stabilité après un oubli (note "Again") — primitive brute.
  ///
  /// Le découpage suit exactement celui de `ts-fsrs` : la borne court terme est
  /// appliquée par l'appelant ([_forgetStabilityClamped]), pas ici. Cela permet
  /// de comparer les primitives une à une entre Dart et TypeScript
  /// (cf. `tools/cross_check.py`, 601 valeurs vérifiées).
  double _nextForgetStability(double d, double s, double r) {
    final double sAfterFail = parameters.w(11) *
        math.pow(d, -parameters.w(12)) *
        (math.pow(s + 1.0, parameters.w(13)) - 1.0) *
        math.exp((1.0 - r) * parameters.w(14));
    return _clamp(sAfterFail, kMinStability, kMaxStability);
  }

  /// Stabilité post-oubli telle qu'utilisée par le planificateur.
  ///
  /// On ne laisse jamais un oubli produire une stabilité supérieure à ce que la
  /// décroissance court terme autorise : oublier une carte ne doit pas pouvoir
  /// allonger son intervalle.
  double _forgetStabilityClamped(double d, double s, double r) {
    final double sShort =
        s / math.exp(parameters.w(17) * parameters.w(18));
    final double upper = _nextForgetStability(d, s, r);
    return _clamp(sShort, kMinStability, upper);
  }

  /// Stabilité pour une revue effectuée le jour même (apprentissage).
  double _nextShortTermStability(double s, Rating rating) {
    final double next = s *
        math.exp(parameters.w(17) * (rating.value - 3 + parameters.w(18)));
    return _clamp(next, kMinStability, kMaxStability);
  }

  // --------------------------------------------------------------------------
  // Planification
  // --------------------------------------------------------------------------

  /// Applique une revue et retourne le nouvel état. Fonction pure.
  ///
  /// [nowMs] est l'instant de la revue ; [cardType] permet de tempérer le gain
  /// de stabilité des QCM (reconnaissance plutôt que rappel actif).
  SrsCardState applyReview(
    SrsCardState current,
    Rating rating,
    int nowMs, {
    CardType cardType = CardType.basic,
  }) {
    final int elapsedDays = current.lastReviewMs == null
        ? 0
        : math.max(0, (nowMs - current.lastReviewMs!) ~/ kMillisPerDay);

    SrsCardState next = current.copyWith(
      elapsedDays: elapsedDays,
      reps: current.reps + 1,
      lastReviewMs: nowMs,
    );

    // --- Première exposition ------------------------------------------------
    if (current.state == CardState.newCard) {
      next = next.copyWith(
        difficulty: _initialDifficulty(rating),
        stability: _initialStability(rating),
      );
      switch (rating) {
        case Rating.again:
          return next.copyWith(
            state: CardState.learning,
            scheduledDays: 0,
            dueMs: nowMs + _minutesToMs(kLearningStepsMinutes[0]),
          );
        case Rating.hard:
          return next.copyWith(
            state: CardState.learning,
            scheduledDays: 0,
            dueMs: nowMs + _minutesToMs(6.0),
          );
        case Rating.good:
          return next.copyWith(
            state: CardState.learning,
            scheduledDays: 0,
            dueMs: nowMs + _minutesToMs(kLearningStepsMinutes[1]),
          );
        case Rating.easy:
          // "Easy" dès la première vue : la carte est déjà connue, on la sort
          // immédiatement de l'apprentissage.
          final int ivl = _clampedInterval(next.stability);
          return next.copyWith(
            state: CardState.review,
            scheduledDays: ivl,
            dueMs: nowMs + ivl * kMillisPerDay,
          );
      }
    }

    final double r = retrievability(elapsedDays.toDouble(), current.stability);
    next = next.copyWith(difficulty: _nextDifficulty(current.difficulty, rating));

    // --- Apprentissage / ré-apprentissage -----------------------------------
    if (current.state == CardState.learning ||
        current.state == CardState.relearning) {
      next = next.copyWith(
        stability: _nextShortTermStability(current.stability, rating),
      );
      switch (rating) {
        case Rating.again:
          return next.copyWith(
            state: current.state,
            scheduledDays: 0,
            dueMs: nowMs + _minutesToMs(kLearningStepsMinutes[0]),
          );
        case Rating.hard:
          return next.copyWith(
            state: current.state,
            scheduledDays: 0,
            dueMs: nowMs + _minutesToMs(10.0),
          );
        case Rating.good:
        case Rating.easy:
          final int ivl = _clampedInterval(next.stability);
          return next.copyWith(
            state: CardState.review,
            scheduledDays: ivl,
            dueMs: nowMs + ivl * kMillisPerDay,
          );
      }
    }

    // --- Révision -----------------------------------------------------------
    //
    // Deux subtilités, toutes deux vérifiées contre `ts-fsrs`
    // (cf. tools/cross_check_sequences.py) :
    //
    // 1. La stabilité se calcule avec la difficulté **d'avant** la revue
    //    (`current.difficulty`), pas avec celle qu'on vient de recalculer.
    //    Utiliser la nouvelle biaiserait systématiquement les intervalles.
    //
    // 2. Une revue le jour même n'est pas un cas particulier : `elapsedDays`
    //    valant 0, R vaut 1 et la formule de rappel ne produit aucun gain.
    //    C'est le comportement voulu — réviser deux fois dans la même journée
    //    n'apprend rien de plus.
    double newStability = rating == Rating.again
        ? _forgetStabilityClamped(current.difficulty, current.stability, r)
        : _nextRecallStability(
            current.difficulty, current.stability, r, rating);

    // Pondération QCM : on conserve l'événement intact mais on réduit le gain.
    if (parameters.enableQcmWeighting &&
        cardType == CardType.qcm &&
        rating != Rating.again &&
        newStability > current.stability) {
      final double gain = newStability - current.stability;
      newStability = _clamp(current.stability + gain * kQcmStabilityWeight,
          kMinStability, kMaxStability);
    }

    next = next.copyWith(stability: newStability);

    if (rating == Rating.again) {
      final int lapses = current.lapses + 1;
      return next.copyWith(
        state: CardState.relearning,
        lapses: lapses,
        scheduledDays: 0,
        dueMs: nowMs + _minutesToMs(kRelearningStepsMinutes[0]),
        isLeech: lapses >= kLeechThreshold,
      );
    }

    final int ivl = _clampedInterval(newStability);
    return next.copyWith(
      state: CardState.review,
      scheduledDays: ivl,
      dueMs: nowMs + ivl * kMillisPerDay,
    );
  }

  /// Les quatre issues possibles, pour afficher l'intervalle sous chaque bouton
  /// **avant** que l'étudiant ne choisisse.
  SchedulingPreview preview(
    SrsCardState current,
    int nowMs, {
    CardType cardType = CardType.basic,
  }) {
    return SchedulingPreview(
      again: applyReview(current, Rating.again, nowMs, cardType: cardType),
      hard: applyReview(current, Rating.hard, nowMs, cardType: cardType),
      good: applyReview(current, Rating.good, nowMs, cardType: cardType),
      easy: applyReview(current, Rating.easy, nowMs, cardType: cardType),
    );
  }

  // --------------------------------------------------------------------------
  // Fold — la règle d'or
  // --------------------------------------------------------------------------

  /// Reconstruit l'état SRS à partir du journal d'événements.
  ///
  /// Propriétés garanties (couvertes par les tests) :
  ///   * **déterminisme** : mêmes événements ⇒ même état, bit à bit ;
  ///   * **indépendance à l'ordre d'insertion** : le tri se fait sur
  ///     `(reviewedAt, id)`, donc deux appareils qui synchronisent dans un
  ///     ordre différent convergent vers le même état ;
  ///   * **idempotence** : les doublons d'identifiant sont éliminés, ce qui
  ///     rend un `push` rejoué inoffensif ;
  ///   * **exclusion du mode examen** : un examen blanc ne décale rien.
  ///
  /// C'est cette fonction qui permet de fusionner deux journaux hors ligne sans
  /// jamais perdre ni inventer de progression.
  SrsCardState fold(Iterable<ReviewEvent> events) {
    final Map<String, ReviewEvent> unique = <String, ReviewEvent>{};
    for (final ReviewEvent e in events) {
      if (e.examMode) continue;
      unique[e.id] = e;
    }

    final List<ReviewEvent> ordered = unique.values.toList()
      ..sort((ReviewEvent a, ReviewEvent b) {
        final int byTime = a.reviewedAtMs.compareTo(b.reviewedAtMs);
        // Départage stable par identifiant : deux revues au même millième de
        // seconde sur deux appareils doivent s'ordonner identiquement partout.
        return byTime != 0 ? byTime : a.id.compareTo(b.id);
      });

    SrsCardState state = SrsCardState.initial;
    for (final ReviewEvent e in ordered) {
      state = applyReview(state, e.rating, e.reviewedAtMs, cardType: e.cardType);
    }
    return state;
  }

  /// Fusionne deux journaux (local et distant) puis rejoue le résultat.
  ///
  /// Utilisé par le moteur de synchronisation (Phase 8) : aucun événement n'est
  /// arbitré ni écrasé, l'union est simplement rejouée.
  SrsCardState mergeAndFold(
    Iterable<ReviewEvent> local,
    Iterable<ReviewEvent> remote,
  ) {
    return fold(<ReviewEvent>[...local, ...remote]);
  }

  // --------------------------------------------------------------------------
  // Utilitaires
  // --------------------------------------------------------------------------

  int _clampedInterval(double stability) {
    final double ivl = intervalFromStability(stability);
    final double rounded = ivl.roundToDouble();
    return _clamp(rounded, 1.0, parameters.maximumInterval.toDouble()).toInt();
  }

  static int _minutesToMs(double minutes) => (minutes * 60000).toInt();

  static double _clamp(double v, double lo, double hi) =>
      v < lo ? lo : (v > hi ? hi : v);
}

File: backend/src/common/fsrs/fsrs.engine.ts
/// FSRS-5 — moteur pur côté serveur.
///
/// Toutes les fonctions sont **pures** : aucun side-effect, aucune lecture
/// d'horloge ou de réseau. `nowMs` est toujours passé en paramètre. C'est
/// la condition de l'équivalence stricte avec le moteur Dart
/// (`mobile/lib/core/srs/fsrs_engine.dart`), vérifiée par les tests.
///
/// Découpage identique à la version Dart :
///   1. primitives (R, intervalle, stabilités, difficulté) ;
///   2. planificateur (applyReview) ;
///   3. fold (règle d'or de la v2 §4, propriété vérifiée par tests).
import {
  CardState,
  CardType,
  FsrsParameters,
  Rating,
  ReviewEvent,
  SchedulingPreview,
  SrsCardState,
  FSRS_DECAY,
  FSRS_FACTOR,
  FSRS_LEARNING_STEPS_MIN,
  FSRS_MAX_DIFFICULTY,
  FSRS_MAX_STABILITY,
  FSRS_MIN_DIFFICULTY,
  FSRS_MIN_STABILITY,
  FSRS_MILLIS_PER_DAY,
  FSRS_QCM_STABILITY_WEIGHT,
  FSRS_RELEARNING_STEPS_MIN,
  DEFAULT_PARAMETERS,
  SRS_INITIAL,
} from './fsrs.constants';

function clamp(v: number, lo: number, hi: number): number {
  return v < lo ? lo : v > hi ? hi : v;
}

function minutesToMs(min: number): number {
  return Math.round(min * 60_000);
}

export class FsrsEngine {
  constructor(private readonly p: FsrsParameters = DEFAULT_PARAMETERS) {}

  // ── Primitives (miroir exact du Dart) ──────────────────────────────────

  retrievability(elapsedDays: number, stability: number): number {
    if (stability <= 0) return 0;
    const t = Math.max(0, elapsedDays);
    return Math.pow(1 + FSRS_FACTOR * (t / stability), FSRS_DECAY);
  }

  intervalFromStability(stability: number): number {
    const r = this.p.requestRetention;
    return (stability / FSRS_FACTOR) * (Math.pow(r, 1 / FSRS_DECAY) - 1);
  }

  private w(i: number): number {
    return this.p.weights[i]!;
  }

  private initialStability(r: Rating): number {
    return clamp(this.w(r - 1), FSRS_MIN_STABILITY, FSRS_MAX_STABILITY);
  }

  private initialDifficulty(r: Rating): number {
    const d = this.w(4) - Math.exp(this.w(5) * (r - 1)) + 1;
    return clamp(d, FSRS_MIN_DIFFICULTY, FSRS_MAX_DIFFICULTY);
  }

  private linearDamping(deltaD: number, difficulty: number): number {
    return (deltaD * (10 - difficulty)) / 9;
  }

  private meanReversion(init: number, current: number): number {
    return this.w(7) * init + (1 - this.w(7)) * current;
  }

  private nextDifficulty(difficulty: number, r: Rating): number {
    const deltaD = -this.w(6) * (r - 3);
    const damped = difficulty + this.linearDamping(deltaD, difficulty);
    const reverted = this.meanReversion(this.initialDifficulty(Rating.Easy), damped);
    return clamp(reverted, FSRS_MIN_DIFFICULTY, FSRS_MAX_DIFFICULTY);
  }

  private nextRecallStability(d: number, s: number, r: number, rating: Rating): number {
    const hardPenalty = rating === Rating.Hard ? this.w(15) : 1;
    const easyBonus = rating === Rating.Easy ? this.w(16) : 1;
    const next = s *
      (1 +
        Math.exp(this.w(8)) *
          (11 - d) *
          Math.pow(s, -this.w(9)) *
          (Math.exp((1 - r) * this.w(10)) - 1) *
          hardPenalty *
          easyBonus);
    return clamp(next, FSRS_MIN_STABILITY, FSRS_MAX_STABILITY);
  }

  private nextForgetStability(d: number, s: number, r: number): number {
    const sAfterFail =
      this.w(11) *
      Math.pow(d, -this.w(12)) *
      (Math.pow(s + 1, this.w(13)) - 1) *
      Math.exp((1 - r) * this.w(14));
    return clamp(sAfterFail, FSRS_MIN_STABILITY, FSRS_MAX_STABILITY);
  }

  private forgetStabilityClamped(d: number, s: number, r: number): number {
    const sShort = s / Math.exp(this.w(17) * this.w(18));
    const upper = this.nextForgetStability(d, s, r);
    return clamp(sShort, FSRS_MIN_STABILITY, upper);
  }

  private nextShortTermStability(s: number, rating: Rating): number {
    const next = s * Math.exp(this.w(17) * (rating - 3 + this.w(18)));
    return clamp(next, FSRS_MIN_STABILITY, FSRS_MAX_STABILITY);
  }

  private clampedInterval(stability: number): number {
    const ivl = this.intervalFromStability(stability);
    const rounded = Math.round(ivl);
    return Math.round(clamp(rounded, 1, this.p.maximumInterval));
  }

  // ── Planificateur ──────────────────────────────────────────────────────

  applyReview(
    current: SrsCardState,
    rating: Rating,
    nowMs: number,
    cardType: CardType = CardType.Basic,
  ): SrsCardState {
    const elapsedDays =
      current.lastReviewMs === null
        ? 0
        : Math.max(0, Math.floor((nowMs - current.lastReviewMs) / FSRS_MILLIS_PER_DAY));

    let next: SrsCardState = {
      ...current,
      elapsedDays,
      reps: current.reps + 1,
      lastReviewMs: nowMs,
    };

    // Première exposition
    if (current.state === CardState.New) {
      next = {
        ...next,
        difficulty: this.initialDifficulty(rating),
        stability: this.initialStability(rating),
      };
      switch (rating) {
        case Rating.Again:
          return {
            ...next,
            state: CardState.Learning,
            scheduledDays: 0,
            dueMs: nowMs + minutesToMs(FSRS_LEARNING_STEPS_MIN[0]!),
          };
        case Rating.Hard:
          return {
            ...next,
            state: CardState.Learning,
            scheduledDays: 0,
            dueMs: nowMs + minutesToMs(6),
          };
        case Rating.Good:
          return {
            ...next,
            state: CardState.Learning,
            scheduledDays: 0,
            dueMs: nowMs + minutesToMs(FSRS_LEARNING_STEPS_MIN[1]!),
          };
        case Rating.Easy: {
          const ivl = this.clampedInterval(next.stability);
          return {
            ...next,
            state: CardState.Review,
            scheduledDays: ivl,
            dueMs: nowMs + ivl * FSRS_MILLIS_PER_DAY,
          };
        }
      }
    }

    const r = this.retrievability(elapsedDays, current.stability);
    next = { ...next, difficulty: this.nextDifficulty(current.difficulty, rating) };

    // Apprentissage / ré-apprentissage
    if (current.state === CardState.Learning || current.state === CardState.Relearning) {
      next = {
        ...next,
        stability: this.nextShortTermStability(current.stability, rating),
      };
      switch (rating) {
        case Rating.Again:
          return {
            ...next,
            state: current.state,
            scheduledDays: 0,
            dueMs: nowMs + minutesToMs(FSRS_LEARNING_STEPS_MIN[0]!),
          };
        case Rating.Hard:
          return {
            ...next,
            state: current.state,
            scheduledDays: 0,
            dueMs: nowMs + minutesToMs(10),
          };
        case Rating.Good:
        case Rating.Easy: {
          const ivl = this.clampedInterval(next.stability);
          return {
            ...next,
            state: CardState.Review,
            scheduledDays: ivl,
            dueMs: nowMs + ivl * FSRS_MILLIS_PER_DAY,
          };
        }
      }
    }

    // Révision (rappel ou oubli)
    let newStability =
      rating === Rating.Again
        ? this.forgetStabilityClamped(current.difficulty, current.stability, r)
        : this.nextRecallStability(current.difficulty, current.stability, r, rating);

    if (
      this.p.enableQcmWeighting &&
      cardType === CardType.Qcm &&
      rating !== Rating.Again &&
      newStability > current.stability
    ) {
      const gain = newStability - current.stability;
      newStability = clamp(
        current.stability + gain * FSRS_QCM_STABILITY_WEIGHT,
        FSRS_MIN_STABILITY,
        FSRS_MAX_STABILITY,
      );
    }

    next = { ...next, stability: newStability };

    if (rating === Rating.Again) {
      const lapses = current.lapses + 1;
      return {
        ...next,
        state: CardState.Relearning,
        lapses,
        scheduledDays: 0,
        dueMs: nowMs + minutesToMs(FSRS_RELEARNING_STEPS_MIN[0]!),
        isLeech: lapses >= FSRS_LEECH_THRESHOLD,
      };
    }

    const ivl = this.clampedInterval(newStability);
    return {
      ...next,
      state: CardState.Review,
      scheduledDays: ivl,
      dueMs: nowMs + ivl * FSRS_MILLIS_PER_DAY,
    };
  }

  preview(current: SrsCardState, nowMs: number, cardType: CardType = CardType.Basic): SchedulingPreview {
    return {
      again: this.applyReview(current, Rating.Again, nowMs, cardType),
      hard: this.applyReview(current, Rating.Hard, nowMs, cardType),
      good: this.applyReview(current, Rating.Good, nowMs, cardType),
      easy: this.applyReview(current, Rating.Easy, nowMs, cardType),
    };
  }

  // ── Fold (la règle d'or) ───────────────────────────────────────────────

  /**
   * Reconstruit l'état SRS à partir d'un journal d'événements.
   *
   * Propriétés garanties (testées) :
   *   * déterminisme : mêmes events ⇒ même état ;
   *   * commutativité : peu importe l'ordre d'insertion ;
   *   * idempotence : doublons d'id ⇒ même état ;
   *   * exclusion du mode examen : un examen blanc ne décale pas.
   */
  fold(events: Iterable<ReviewEvent>): SrsCardState {
    const unique = new Map<string, ReviewEvent>();
    for (const e of events) {
      if (e.examMode) continue;
      unique.set(e.id, e);
    }
    const ordered = Array.from(unique.values()).sort((a, b) => {
      const byTime = a.reviewedAtMs - b.reviewedAtMs;
      return byTime !== 0 ? byTime : a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
    });
    let state: SrsCardState = SRS_INITIAL;
    for (const e of ordered) {
      state = this.applyReview(state, e.rating, e.reviewedAtMs, e.cardType);
    }
    return state;
  }

  mergeAndFold(local: Iterable<ReviewEvent>, remote: Iterable<ReviewEvent>): SrsCardState {
    return this.fold([...local, ...remote]);
  }
}
